src/main/java/edu/wpi/grip/core/Step.java

```diff
@@ -78,13 +79,50 @@ protected Pipeline getPipeline() {
         return this.pipeline;
     }
 
+    /**
+     * Resets all {@link OutputSocket OutputSockets} to their initial value.
+     * Should only be used by {@link Step#runPerformIfPossible()}
+     */
+    private void resetOutputSockets() {
+        for (OutputSocket<?> outputSocket : outputSockets) {
+            outputSocket.resetValueToInitial();
+        }
+    }
+
+    /**
+     * The {@link Operation#perform} method should only be called if all {@link InputSocket#getValue()} are not empty.
+     * If one input is invalid then the perform method will not run and all output sockets will be assigned to their
+     * default values.
+     */
+    private synchronized void runPerformIfPossible() {
+        try {
```

Comment:
This could be rewritten like this to be much clearer and not use exceptions for control flow:

    for (InputSocket<?> socket : this.inputSockets) {
        if (!socket.getValue().isPresent()) {
            this.resetOutputSockets();
            return;
        }
    }